manager/volume.go

```diff
@@ -614,7 +617,59 @@ func (m *VolumeManager) Expand(volumeName string, size int64) (v *longhorn.Volum
 	}
 
 	if v.Spec.Size > size {
-		return nil, fmt.Errorf("cannot expand volume %v with current size %v to a smaller size %v", v.Name, v.Spec.Size, size)
+		return nil, fmt.Errorf("longhorn volume %v spec size %v is greater than requested Persistent Volume Claim size %v", v.Name, v.Spec.Size, size)
+
+	}
+
+	if v.Status.ActualSize > size {
+		return nil, fmt.Errorf("longhorn volume %v actual size %v is greater than requested Persistent Volume Claim size %v", v.Name, v.Spec.Size, size)
+	}
+
+	kubernetesStatus := &v.Status.KubernetesStatus
+	if kubernetesStatus.PVCName != "" && kubernetesStatus.LastPVCRefAt == "" {
+		pvc, err := m.ds.GetPersistentVolumeClaim(kubernetesStatus.Namespace, kubernetesStatus.PVCName)
+		if err != nil {
+			return nil, err
+		}
+
+		requestedSize := resource.MustParse(strconv.FormatInt(size, 10))
+
+		// TODO: Should check for pvc.Spec.Resources.Requests.Storage() here, once upgrade API to v0.18.x.
+		storageSpecValue, ok := pvc.Spec.Resources.Requests[corev1.ResourceStorage]
+		if !ok {
+			return nil, fmt.Errorf("cannot get request storage")
+		}
+		// TODO: Should check for pvc.Status.CapacityStorage() here, once upgrade API to v0.18.x.
+		storageStatusValue, ok := pvc.Status.Capacity[corev1.ResourceStorage]
+		if !ok {
+			return nil, fmt.Errorf("cannot get capacity storage")
+		}
+
+		if storageSpecValue.Cmp(requestedSize) == 0 && storageStatusValue.Cmp(requestedSize) == 0 && v.Spec.Size == size {
+			logrus.Infof("Persistent Volume Claim %v expansion is not necessary since current size %v == %v", v.Name, v.Spec.Size, size)
+			return v, nil
+		} else if storageSpecValue.Cmp(requestedSize) > 0 {
+			return nil, fmt.Errorf("persistent volume claim %v spec size %v is greater than requested persistent volume claim size %v", kubernetesStatus.PVCName, storageSpecValue.Value(), requestedSize.Value())
+		} else if storageStatusValue.Cmp(requestedSize) > 0 {
```

Comment:
Why we ned to check status value?

Reply:
if user manually created pvc, we can see request size 1Gi and status size is 2Gi

Reply:
If users request 1G, but the existing size is bigger than it. The following logic should be handled by the Longhorn CSI plugin. 

Yes, it's good to tackle here, but actually, the backend is also taking care of. I am fine with no need to change. 

https://github.com/longhorn/longhorn-manager/blob/713e2e48b6f7fe191ccd36d33b41fb2e5a234811/csi/controller_server.go#L672-L682